### ISSUE TITLE
Fix false positives for inlines styles in declaration-empty-line-before

### ISSUE THREAD

**Issue by WangHansen** [stylelint/stylelint#4658]
> Clearly describe the bug

`declaration-empty-line-before` rule force a new line to be inserted for inline style

> Which rule, if any, is the bug related to?

`declaration-empty-line-before`

> What code is needed to reproduce the bug?

e.g.

```html
<span
  style="
    font-size: 16px;
    color: black;
    font-weight: bold;
  "
>
  Text
</span>
```

> What stylelint configuration is needed to reproduce the bug?

This is from the [stylelint-config-standard](https://github.com/stylelint/stylelint-config-standard/blob/f783c35fd4632430b00fa8169307bb03a62ee058/index.js#L50) config

```json
"declaration-empty-line-before": ["always", {
   "except": [
      "after-declaration",
      "first-nested",
   ],
   "ignore": [
      "after-comment",
      "inside-single-line-block",
   ],
} ],
```

> Which version of stylelint are you using?

`13.2.1`

> How are you running stylelint: CLI, PostCSS plugin, Node.js API?

vscode plugin and CLI with `stylelint --color **/*.{scss,vue}`

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

No, just inline-style and multiple lines inline-style

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors did you get)?

The following warnings were flagged:
![image](https://user-images.githubusercontent.com/15679461/77242424-9da07980-6bd4-11ea-9c78-7e897cc579a6.png)

**Comment by jeddy3** [comment]
@WangHansen Thanks for the report and for using the template.

I can reproduce this [using the demo](https://stylelint.io/demo#N4Igxg9gJgpiBcIA8BnADgQwHYB0sAJ8UAXATwBsYBeHEPQwgMwi2IFoUBLALxnnwCMANjQAPANz0GkchABO-AEbkMYANaSCDZqzYB3GJwDmAC2JKI5KJsK08APikAVGKOJ4kAenTZ7IADQgKKSsGKIIIGYAtuQB4CyMxhHAUrRyAK6UKLT8KVq2IBjsGZRsWBBs6Vhq5Xq4CPjEGTD+UgXKEOplFTBRaGQ5jc2t+fi0MvLdbJxYAG4Y5JxQbCaug03pLW1j8VFRMLrlbL39pOvD27SwYCpyRZwsbB1dR1DpaItgRTBsaHIQaBgcmInBg2QaAG1toQNltRoQ8gwkTtjOU5DBBlD4cjxiwUDAwOkQbMfm8PpwvsQwfpOMQTGwoJxGIwgQd2PNyJtwdCGABdHkAX22vJGSKuBNu90ezzUUxQJnkdOwyz+AKBZDYEBJcjkSzB502ooYtB07EYGCinHIpCmZM+3zKFv1DVhRoKprY5st1qmlpQXCwRjYRgOQIpnotVptahgpD08igBrhYpAjCqYBBjy+5DAUxmHKWSbdOzTWAzDywbGzuaOVR8YBgyzVd2I8iLl1T6czlcWWBgGDkwbujLZU3KWBIyoHy0Z6PLLHbo1oMdIjDu+wZEoHUsrR04fUV2HcLouS5A+0ZGE9-eI6XRjo3teqtXq-FdHdeYIbWEZgY4gLAJkKVpM4T0ND8KjtCkHSKaYD2BNgSmdN9TxTV53ntKkOBgSgM3kcEUPAs8jhODUUAgO8G0XNCejcO5sMtCYJ2o40QD3OYFiWBkKOUH4UBUeUqwgPY2QIoYiJTVVAWBG0nxqCA6hYgp8Vw1tBzQfF0igCobgwf0piqeTFLA5NWJUgk1N+TTtOOSh9kOCpDJfJSdnMvDBzIQEDOfBTX3E0zlKaGYgyOPs9F7DETOLWgqlpbyjL898z1gpDjj6DUIqeGBmHRTEeVoBY9AwUhwWLBEeQKVwG36PLsRTGUshQNgMEYKlBxQJ0Hx+BqwVKirWMSOQSDKMEqUTOg6vwEV+toVF5Ei-gsWRHFClaoEhJE1g7Em-lsSFUZpuS4pMh+LBOq+fFBloWQDDkWhosKY7SjOjd6x+Fq2quwpyCKkqOGC0oIvujsUpOhiKUsR4wsyj6gS+wrir6s8ZSrWQAyDRQ7gbNLTjYTLFGy+avr7bVgeR2QuhuCB0aeLHTpgcKZneta7oaAqfsRsn6op2UqZpzHVHpxm+yynKFp2BG-qiTIQTxpmudYlG+eC2nBf-NWCbF+GOb+9HAflgIOxRtUsBVgXsehpnmpZ7Xfqa6XyFloHDfJzpZRNs26fV7HYdZ-h2bt-7A31vsFfaHnNUBU2-3NvjMGxzWibZ767bDnYJkHVZRCrPTxeuhS4ZdlMM5WVw8YOIw6S++VFTT3FNvYUibXxwncshfKU85ovkURZadiqmAavbyb3U4Ib2D7EhG22vuprK1i5rbxb+sCigcJmdhID2ZVuR2wVhQerf7PYPQTFpMF45+GYuFgW2u8P9ISGE35-mkjUm7lkXE6X-AlpWyWkbLV7stWgA8h7LxHhLFmVZH6tiiC-ZsAxu6zxNGPYak8xozz7odFBIBF7iz-rg32G1j5pxAXgiceog5GBDt1HmWDlq7WWvtJEODWLXElN2J42AgxvWtp9ZOJNC4PQ4duLhigeHe26q3POncSp1xAKIlsFYngR3xIxSGu4GYwxtsnABbAHZOwNiIrcyjpRqN6BDWQlY+G+zvrrAGPxnYmN0mYysKN1FWMeHw7+sihGsxcZwlRHjHGfx+FEMIm5XE7jEgIQJYjgkRyaBgK0KtPFMXsYA9hpidxCWsSNYWzMBH+zkfbGWnAwkKKUbkpiUj+GFxKfovWTjjEdmqVw2pPiZHExgKTZBOx2kqI-i3LWw9yEALIfgYBK0wHHggbPAoxDBkLn6eQwa6DRrTwmrPNh5D8G1QWVAtqJC2STJTNfKhzSwmqLdgw5ETDkQsL5A9Us85KxHyvJbEWdi9E6zKY7CpziOyvI6cJCJdSfmNL+dQ2hCiQUqI+XU3xPS+kvK7CoiJ2dhlMzEgABjRWWLhL0fgXVkTdYRwL0WPEwOiVgqx8RNS+VfSht9fmB0MYC1pZ54XUoHGyel1I+EXNZSU-xMKWmh36SaKllZT7n1sboqFqcpUUNgKEbsgwABMD1KCBjpGwXg-xvK0hcrQTFOMMo4sGHEjsF5OBXhZEUO8JLNEQsVRLP5Ci7UOpvM6vJ3jL6iyTqK3pFKzzeuvE6+8xKc6XWTuSgJtrGz2sjbee8NL+VgkFYG4VfjQ2JvDcmn1Uafh3EDD8ZsRRJgKuKR65VD0I2OrTaWnhFbpJVvaoG5FbL75JsvGwAAjpsOQzdODDURUy+pfs62IwMeUypKqI1DqBKO8dYKrw1oaTOhxwcJWRQbUWwdw7V2b3XUi7pgj80KJIpYfQZ8qRvVNWxCofoaYHCbIwDgFE5AW20QbQiAUdhYHSFEAmg5KAYF-EGQ1EBMnXpA2BqYyTUl-hg2JJKklX7qhtKSr6CaFGpWxV-C98z-7Qo5Qu+e0yUyzIOQs1B48RpT3Gv1XZK19ljPo6tY5R9TnbOwfvA6D03KWSKEFRQRJupYxjMQJqQqWV5tRR2ETkwxO6gk1hStllN3TtoP4hRKnBxqc4BpttQIO3ntGSGpTZ5DMbUUDMCzOm4Mqrs1vBzZ1tNdpI9urJykcIWUmLAFA34oBHns45yyoUA2CyfXZxYa7t4FJ0bWgOvbbMBfcnLRL4KulWaA1e1zmXLLmqI8hfA+LlPFcmBpGAWkdICSarh+NBcC0pjs7V+rqM9JNQzXSrNcmc0KZRWG9r1X1LWR6HZUczWSn4aK6pGrk3bK9Bm668oxAdxfW0ukXiBnxtsE8iS3OeHWsKNipvE7LXboKI5JsNgK54xyGWLNnY82Hp3ZaTlz5f7vnurS1LedQKzyfey6epLzme07poXu27Cx7sJfB7l7z+W9OFY+-Dr77BSvpVHZPQYuKpD7QFCAAUQA).

This rule should, like the `at-rule-empty-line-before` and `rule-empty-line-before`, ignore nodes that are the first ones of root. The fix should be straightforward: use the [`isFirstNodeOfRoot` util](https://github.com/stylelint/stylelint/blob/master/lib/utils/isFirstNodeOfRoot.js) to return early if this is the case, [like we do in `at-rule-empty-line-before`](https://github.com/stylelint/stylelint/blob/master/lib/rules/at-rule-empty-line-before/index.js#L65-L68).

I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to fix a bug in a rule](https://stylelint.io/developer-guide/rules#fix-a-bug-in-a-rule) in the Developer guide.

**Comment by malsf21** [comment]
If nobody else is doing this, I can do this issue sometime soon (probably during the weekend). Just to confirm, I'll add the `isFirstNodeOfRoot` check and probably try to add a test to confirm the behaviour.

**Comment by jeddy3** [comment]
> If nobody else is doing this, I can do this issue sometime soon

Excellent. I'll label as "wip".

> I'll add the isFirstNodeOfRoot check and probably try to add a test to confirm the behaviour.

That's correct. You can test HTML code by using the `syntax` property of the `testRule` schema. Like this rough cut (indentation and test correctness likely off):

```
testRule(rule, {
	ruleName,
	config: ["always],
	syntax: 'html',

	accept: [
		{
			code: `<span style="color: red;"></span>`,
			description: 'Singleline HTML style tag',
		},
                 {
			code: `<span style="color: red; font-size: 1rem;"></span>`,
			description: 'Singleline HTML style attribute with two declarations',
		},
		{
			code: stripIndent`<span
                              style="
                                 color: red;
                                 font-size: 1rem;
                             "></span>`,
			description: 'Multiline HTML style attribute with two declarations',
		},
          ]
          reject: [
               {
			code: stripIndent`<span style="color: red;"></span>
                            <style>
                               color: red;
                               font-size: 16px;
                            <style>`,
			description: 'flush declaration in style tag',
			message: messages.rejected,
			line: 4,
			column: 8,
		}
         ]
     }
});
```

You can put your tests at the bottom of the test file.

**Comment by malsf21** [comment]
Hi @jeddy3, thank you for the help so far! I implemented the check & added the suggested tests (with a few syntax and indentation fixes), but it doesn't seem like it still passes the tests for the issue's original concern (inline style declarations). In particular, I'm not sure if `isFirstNodeOfRoot` is enough - should I instead implement something that ignores the rule completely if it's an inline style declaration? Or do I not understand the rule correctly?

Attached is also the error log, though I understand it might be a bit tricky to sift through. I noticed that the errors occur on the second declaration/attribute set, which seems in indicate that `isFirstNodeOfRoot` is working but something else is causing the problem.

You can also view the changes on the branch [`false-positive-declaration-empty-line-before`](https://github.com/malsf21/stylelint/tree/fix/false-positive-declaration-empty-line-before) of my fork.

```
  ● declaration-empty-line-before › accept › [ 'always' ] › '<span style="color: red; font-size: 1rem;"></span>' › Single-line HTML style attribute with two declarations

    expect(received).toEqual(expected) // deep equality

    - Expected  - 1
    + Received  + 9

    - Array []
    + Array [
    +   Object {
    +     "column": 26,
    +     "line": 1,
    +     "rule": "declaration-empty-line-before",
    +     "severity": "error",
    +     "text": "Expected empty line before declaration (declaration-empty-line-before)",
    +   },
    + ]

      34 | 				const output = await stylelint(options);
      35 | 
    > 36 | 				expect(output.results[0].warnings).toEqual([]);
         | 				                                   ^
      37 | 				expect(output.results[0].parseErrors).toEqual([]);
      38 | 
      39 | 				if (!schema.fix) return;

      at Object.<anonymous> (jest-setup.js:36:40)

  ● declaration-empty-line-before › accept › [ 'always' ] › '<span\n\tstyle="\n\t\tcolor: red;\n\t\tfont-size: 1rem;\n\t">\n\tText\n</span>' › Multi-line HTML style attribute with two declarations

    expect(received).toEqual(expected) // deep equality

    - Expected  - 1
    + Received  + 9

    - Array []
    + Array [
    +   Object {
    +     "column": 3,
    +     "line": 4,
    +     "rule": "declaration-empty-line-before",
    +     "severity": "error",
    +     "text": "Expected empty line before declaration (declaration-empty-line-before)",
    +   },
    + ]

      34 | 				const output = await stylelint(options);
      35 | 
    > 36 | 				expect(output.results[0].warnings).toEqual([]);
         | 				                                   ^
      37 | 				expect(output.results[0].parseErrors).toEqual([]);
      38 | 
      39 | 				if (!schema.fix) return;

      at Object.<anonymous> (jest-setup.js:36:40)

```

**Comment by jeddy3** [comment]
@malsf21 Thanks for getting so far! You are almost there. The tests just needed tweaking as the rule was behaving as expected:

```js

testRule(rule, {
	ruleName,
	config: ['always'],
	syntax: 'html',
	fix: true,

	accept: [
		{
			code: `<span style="color: red;"></span>`,
			description: 'Single-line HTML style tag',
		},
		{
			code: stripIndent`
				<span
					style="
						color: red;

						font-size: 1rem;
					">
					Text
				</span>`,
			description: 'Multi-line HTML style attribute with two declarations',
		},
	],
	reject: [
		{
			code: stripIndent`
				<span style="color: red;font-size: 16px;"></span>
				<style>
					color: red;
					font-size: 16px;
				</style>`,
			fixed: stripIndent`
				<span style="color: red;

				font-size: 16px;"></span>
				<style>
					color: red;

					font-size: 16px;
				</style>`,
			description: 'flush declaration in style tag',
			warnings: [
				{
					message: messages.expected,
					line: 1,
					column: 25,
				},
				{
					message: messages.expected,
					line: 4,
					column: 2,
				},
			],
		},
	],
});

testRule(rule, {
	ruleName,
	config: ['always', { ignore: ['inside-single-line-block'] }],
	syntax: 'html',
	fix: true,

	accept: [
		{
			code: `<span style="color: red; font-size: 1rem;"></span>`,
			description: 'Single-line HTML style attribute with two declarations',
		},
		{
			code: stripIndent`
				<span
					style="
						color: red;

						font-size: 1rem;
					">
					Text
				</span>`,
			description: 'Multi-line HTML style attribute with two declarations',
		},
	],
	reject: [
		{
			code: stripIndent`
				<span style="color: red; font-size: 1rem;"></span>
				<style>
					color: red;
					font-size: 16px;
				</style>`,
			fixed: stripIndent`
				<span style="color: red; font-size: 1rem;"></span>
				<style>
					color: red;

					font-size: 16px;
				</style>`,
			description: 'flush declaration in style tag',
			message: messages.expected,
			line: 4,
			column: 2,
		},
	],
});
```

> should I instead implement something that ignores the rule completely if it's an inline style declaration,

We don't want to ignore inline style declarations outright because users may want to enforce an empty line code style for them, like so:

```html
<span style="
  color: red;
  font-size: 16px;"
>Text</span>

<style>
  color: red;
  font-size: 16px;
</style>
```

In the tests, `<span style="color: red; font-size: 1rem;"></span>` was being flagged because with the `"always"` option it is expecting:

```html
<span style="color: red; 

font-size: 1rem;"></span>
```

Users can ignore single line declaration lists using the `ignore: ['inside-single-line-block'] }` secondary option (with the `"always"` primary option). I added some tests to show this.

For example, with:

```json
{
  "rules": {
    "declaration-empty-line-before": [
      "always",
      {
        "except": ["after-declaration"],
        "ignore": ["inside-single-line-block"],
      },
    ]
  }
}
```

The following code styles are all valid.

```html
<span style="
  color: red;
  font-size: 16px;"
>Text</span>

<style>
  color: red;
  font-size: 16px;
</style>

<span style="color: red; font-size: 16px;">Text</span>
<style>color: red; font-size: 16px;</style>
```

But the following are not:

```html
<span style="
  color: red;

  font-size: 16px;"
>Text</span>

<style>
  color: red;

  font-size: 16px;
</style>
```

I hope that clarifies things for you. Thanks again for implementing the changes. Do you want to copy my tests across to your fork and then submit a pull request?

**Comment by malsf21** [comment]
Thank you very much for the detailed explanation @jeddy3, I think I was confused with how the rules interacted but the examples were very helpful. I'll adjust the tests to my fork and submit a PR tonight. Thank you again for the help!
